Remove dry-configurable dependency

GustavoCaso opened the pull request:
While working on the new version of `dry-configurable` https://github.com/dry-rb/dry-configurable/pull/49 we decided to use `dry-struct` to store the config information, but we found that using `dry-struct` would be possible, because it will create a circular dependency: 

`dry-struct` requires `dry-types` which requires `dry-container` which require `dry-configurable` 💥 

![Dependency Graph](https://files.gitter.im/dry-rb/core-team/XnlQ/gem_graph.png)

For the issue to be resolve, we decided that the best approach would be to remove `dry-configurable` from here and use [dry-core Class Attributes](https://github.com/dry-rb/dry-core/blob/master/lib/dry/core/class_attributes.rb) to store both `registry`, `resolver` and `namespace_separator`

GustavoCaso committed:
- Remove dry-configurable dependency
- Use dry-core class attributes module

GustavoCaso committed:
Update ruby and jruby on travis

timriley commented on lib/dry/container/mixin.rb:
```diff
@@ -235,6 +239,18 @@ def import(namespace)
+
+      def _namespace_separator
+        respond_to?(:namespace_separator) ? namespace_separator : self.class.namespace_separator
+      end
```
@GustavoCaso What's the thinking behind these `respond_to?` checks, rather than just delegating directly to `self.class`?

GustavoCaso replied:
I check using `respond_to?` because since `dry-container` could be used at class level and instance level, when the caller is at class level we get this error `undefined method `resolver' for Class:Class`

I could check instead of using `respond_to?` I could check whether is a class or an instance.

WDYT? @timriley @solnic

timriley replied:
@GustavoCaso Ah yeah, I see the issue now.

My first thought was that maybe we should just "bless" a single type of usage for the dry-container mixin – either include or extend – but I think there are valid use cases for both :)

So, if instead we just want to find a way for these internal methods to read a little nicer, how about something like this?

```ruby
def namespace_separator
  defined?(super) ? super : self.class.namespace_separator
end
```

By using the `defined?(super)` check and calling super if it's there, it means at least we no longer need to prefix these method names with an underscore.

Another approach could be to invert the check:

```ruby
def namespace_separator
  self.class.respond_to?(:namespace_separator) ? self.class.namespace_separator : super
end
```

Whichever approach we take, I think it might be good to prefix all these similarly-behaving methods with a comment explaining why they work like they do.

solnic replied:
We should make it in a way that there's no need for any checks at run-time. Which means a container class shouldn't have these methods at all, as they are provided by `ClassAttributes`, and a container object should have them implemented and they would simply delegate to `self.class`.

GustavoCaso replied:
@timriley Thank you so much for the answer.

I never use or thought about the possibility of `defined?(super)` but after your explanation makes a lot of sense. For clarity reason I would stick with the second approach I think it reads better for these case.

Also totally agree with the comment to also help understand the reason behind them.

timriley replied:
@GustavoCaso Nice outcome in https://github.com/dry-rb/dry-container/pull/41/commits/ee0877a0e09d927f72f4b45f48b161fca3b73e92 :)

solnic commented on lib/dry/container/mixin.rb:
```diff
@@ -235,6 +239,18 @@ def import(namespace)
+
+      def _registry
+        respond_to?(:registry) ? registry : self.class.registry
+      end
```
what's the reason for checking `respond_to?`

GustavoCaso committed:
Remove the need for checking at run-time if the methods resolver, resgisty and
namespace_separator and defined at the class level or the instance
level.

- At the class level are defined by Dry::Core::ClassAttributes
- At the Instance level we implement them and simply delegate to the
  class level methods

GustavoCaso commented:
@solnic @timriley Thanks for the comments. I updated the code following the suggestions.

solnic commented on dry-container.gemspec:
```diff
@@ -18,7 +18,7 @@ Gem::Specification.new do |spec|
 
   spec.add_runtime_dependency 'concurrent-ruby', '~> 1.0'
-  spec.add_runtime_dependency 'dry-configurable', '~> 0.1', '>= 0.1.3'
+  spec.add_runtime_dependency 'dry-core', '~> 0.4.2'
```
please relax this a bit to `~> 0.4`

GustavoCaso committed:
relax version for dry-core
